peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+other than ``property`` are rejected.
+
+Covering the extra possibilities induces a great amount of boilerplate, involving
+creation of an abstract descriptor protocol, possibly also accounting for
+class and instance level overloads.
+Worse yet, all of that is multiplied for each additional read-only attribute.
+
+
+Rationale
+=========
+
+These problems can be resolved by an attribute-level :external+typing:term:`type qualifier`.
+``ReadOnly`` has been chosen for this role, as its name conveys the intent well,
+and the newly proposed changes complement its semantics defined in the :pep:`705`.
+
+A class with a read-only instance attribute can be now defined as such::
+
+    from typing import ReadOnly
+
+
+    class Member:
+        def __init__(self, id: int) -> None:
+            self.id: ReadOnly = id
+
+...and a protocol described in :ref:`protocols` is now just::
+
+    from typing import Protocol, ReadOnly
+
+
+    class HasName(Protocol):
+        name: ReadOnly[str]
+
+
+    def greet(obj: HasName, /) -> str:
+        return f"Hello, {obj.name}!"
+
+* A subclass of ``Member`` can redefine ``.id`` as a writable attribute or a
+  :term:`descriptor`. It can also :external+typing:term:`narrow` the type.
+* The ``HasName`` protocol can be implemented by any mechanism allowing for ``.name`` access.
+* The ``greet`` function can now accept a wide variety of compatible objects,
+  while being explicit about no modifications being done to the input.
+
+
+Specification
+=============
+
+The :external+py3.13:data:`typing.ReadOnly` :external+typing:term:`type qualifier`
+becomes a valid annotation for :term:`attributes <attribute>` of classes and protocols.
+It is used to indicate that an attribute should not be reassigned or ``del``\ eted.
+
+The deletability rule should be extended to ``Final`` as well, as it is currently
+not specified.
+
+Akin to ``Final``, read-only attributes do not influence the mutability of
+the assigned object. Immutable ABCs and containers may be used in combination with
+``ReadOnly`` to prevent mutation of such values.
+
+Syntax
+------
+
+``ReadOnly`` can be used at class-level or within ``__init__`` to mark individual
+attributes read-only:
+
+.. code-block:: python
+
+    class Book:
+        id: ReadOnly[int]
+
+        def __init__(self, id: int, name: str) -> None:
+            self.id = id
+            self.name: ReadOnly = name
+
+The explicit type in ``ReadOnly[<type>]`` can be omitted if the declaration has
+an initializing value. A type checker should apply its usual type inference
+rules to determine the type of ``name``.
+
+If an attribute is already implied to be read-only, like in the frozen :ref:`classes`,
+explicit declarations should be permitted and seen as equivalent, except ``Final``
+additionally forbids overriding in subclasses:
+
+.. code-block:: python
+
+    @dataclass(frozen=True)
+    class Point:
+        x: ReadOnly[int]
+        y: Final[int]
```

Comment:
Could use an example where a child class overrides the ReadOnly field with a narrower type. Also, might be worth commenting on the fact that ReadOnly is redundant here, as all fields are implicitly read-only.